media_driver/linux/common/os/i915/mos_bufmgr.c

```diff
@@ -4631,3 +4631,132 @@ mos_bufmgr_gem_init(int fd, int batch_size)
+            strerror(errno));
+        free(context);
+        return nullptr;
+    }
+
+    context->ctx_id = create.ctx_id;
+    context->bufmgr = bufmgr;
+
+    return context;
+}
+
+
+int mos_query_engines(int fd,
+                      __u16 engine_class,
+                      __u64 caps,
+                      unsigned int *nengine,
+                      struct class_instance *ci)
+{
+    struct drm_i915_query query;
+    struct drm_i915_query_item query_item;
+    struct drm_i915_query_engine_info *engines;
+    int ret, len;
+
+    engines = (drm_i915_query_engine_info *)malloc(4096);
+    memset(engines,0,4096);
+    memclear(query_item);
+    query_item.query_id = DRM_I915_QUERY_ENGINE_INFO;
+    query_item.length = 0;
+    query_item.data_ptr = (uintptr_t)engines;
+    memclear(query);
+    query.num_items = 1;
+    query.items_ptr = (uintptr_t)&query_item;
+
+    ret = drmIoctl(fd, DRM_IOCTL_I915_QUERY, &query);
+    if (ret)
+    {
+        goto fini;
+    }
+    len = query_item.length;
+
+    memset(engines,0,4096);
+    memclear(query_item);
+    query_item.query_id = DRM_I915_QUERY_ENGINE_INFO;
+    query_item.length = len;
+    query_item.data_ptr = (uintptr_t)engines;
+    memclear(query);
+    query.num_items = 1;
+    query.items_ptr = (uintptr_t)&query_item;
+
+    ret = drmIoctl(fd, DRM_IOCTL_I915_QUERY, &query);
+    if (ret)
+    {
+        goto fini;
+    }
+
+    int i, num;
+    for (i = 0, num = 0; i < engines->num_engines; i++) {
+        struct drm_i915_engine_info *engine =
+            (struct drm_i915_engine_info *)&engines->engines[i];
+        if ( engine_class == engine->engine_class
+             && ((caps & engine->capabilities) || caps == 0))
+        {
+            ci->engine_class = engine_class;
+            ci->instance = engine->instance;
+            ci++;
+            num++;
+        }
+    }
+    if (num > *nengine)
+    {
+        fprintf(stderr,"%s: Number of engine instances out of range, %d,%d\n",
+                __FUNCTION__, num, *nengine);
+        goto fini;
+    }
+    *nengine = num;
+
+
+fini:
+    free(engines);
+    return ret;
+}
+
+int mos_set_context_param_load_balance(struct mos_linux_context *ctx,
+                     const struct class_instance *ci,
+                     unsigned int count)
+{
+    struct balancer {
+        uint64_t next_extension;
+        uint64_t name;
+
+        uint64_t flags;
+        uint64_t mask;
+
+        uint64_t mbz[4];
+    } balancer;
+    memclear(balancer);
+    balancer.name  = I915_CONTEXT_ENGINES_EXT_LOAD_BALANCE;
+    balancer.mask = ~0ull;
+
+    struct set_engines {
+        uint64_t extension;
+        struct class_instance engines[MAX_VDBOX_NUM];
+    } set_engines;
+    set_engines.extension = (uintptr_t)(&balancer);
+    memcpy(set_engines.engines, ci, count * sizeof(struct class_instance));
+    return mos_set_context_param(ctx,
+                          sizeof(set_engines.extension) + count*sizeof(struct class_instance),
```

Comment:
how count relates to MAX_VDBOX_NUM? Shouldn't we have buffer overflow protection here?

Reply:
count is returned from KMD which should always <= the array size MAX_VDBOX_NUM. But yet, should add protection here.

Reply:
MAX_VDBOX_NUM should not be used here. Will revise the patch.

Reply:
done.